### PR TITLE
Updating data provider docs

### PR DESCRIPTION
**Pull request by robertbastian**
#2945

### CODE REVIEW

**Comment by robertbastian** [comment]
http://icu4x-pr-artifacts.storage.googleapis.com/gha/656848a9aadf58a02534408d2edf7c8c944d9ee9/docs/icu/index.html

File: components/icu/src/lib.rs [hunk 1]
```diff
@@ -18,83 +18,102 @@
...
 //! )
-//! .expect("Failed to create DateTimeFormatter instance.");
+//! .expect("compiled data should include 'es-US'");
+//! ```
+//!
+//! Using compiled data data has several benefits:
+//! * Simple code, no need to manage [`DataProvider`]s
+//! * No deserialization, improves binary size and performance
+//! * The compiler can discard data for unused APIs. As ICU4X's APIs are very fine-grained, you only
+//!   have to carry a minimal amount of data
```
Comment (sffc) [inline components/icu/src/lib.rs:44]: "several benefits" sounds like we're pitting compiled data and explicit data against each other, when both of them can have no deserialization, improved binary size, etc. How about

```suggestion
//! Clients using compiled data benefit from simple code and optimal zero-cost data loading. Additionally,
//! ICU4X's APIs are designed such that dead-code elimination can optimize away unused compiled data.
```

File: components/icu/src/lib.rs [hunk 2]
```diff
@@ -18,83 +18,100 @@
...
-//!     options,
+//!     &locale!("es-US").into(),
+//!     Default::default(),
 //! )
-//! .expect("Failed to create DateTimeFormatter instance.");
+//! .expect("compiled data should include 'es-US'");
+//! ```
+//!
+//! Clients using compiled data benefit from simple code and optimal zero-cost data loading. Additionally,
+//! ICU4X's APIs are designed such that dead-code elimination can optimize away unused compiled data.
```
Comment (Manishearth) [inline components/icu/src/lib.rs:42]: nit: perhaps mention that this is per data type not per locale

**Review by Manishearth — APPROVED** [review]

File: components/icu/src/lib.rs (hunk 2 above)
Comment (robertbastian) [reply components/icu/src/lib.rs:42]: I don't really want to get into defining what a data key is here.

**Review by Manishearth — APPROVED** [review]

**Pull request merged**